components/Buttons/src/MDCButton.m

```diff
@@ -890,21 +890,32 @@ - (void)updateTitleFont {
 
   if (_mdc_adjustsFontForContentSizeCategory) {
     // Dynamic type is enabled so apply scaling
-    if (font.mdc_scalingCurve) {
-      font = [font mdc_scaledFontForTraitEnvironment:self];
-    } else {
-      if (self.adjustsFontForContentSizeCategoryWhenScaledFontIsUnavailable) {
-        font = [font mdc_fontSizedForMaterialTextStyle:MDCFontTextStyleButton
-                                  scaledForDynamicType:YES];
-      }
-    }
+    font = [self updateFontForDynamicTypeWithFont:font];
   }
 
   self.titleLabel.font = font;
 
   [self setNeedsLayout];
 }
 
+- (UIFont *)updateFontForDynamicTypeWithFont:(UIFont *)font {
+  if (font.mdc_scalingCurve) {
+    font = [font mdc_scaledFontForTraitEnvironment:self];
+  } else {
+    if (self.adjustsFontForContentSizeCategoryWhenScaledFontIsUnavailable) {
+      font = [font mdc_fontSizedForMaterialTextStyle:MDCFontTextStyleButton
+                                scaledForDynamicType:YES];
+    }
+  }
+  // Update the _font dictionary so that it returns the correct font.
+  if (_fonts[@(self.state)] != nil) {
+    _fonts[@(self.state)] = font;
+  } else {
+    _fonts[@(UIControlStateNormal)] = font;
+  }
```

Comment:
Is this the expected behavior in terms of UIKit? If so, please add a test to this suite that verifies the UIKit behavior if possible so that we can catch any changes in UIKit's implementation over time.

Reply:
This is a deviation from UIKit but the expected behavior within our components.